### PR TITLE
Backup and restore commands

### PR DESCRIPTION
**Pull request by jakedt**
This consumes the experimental bulk import and export APIs from SpiceDB, which were added to the API in [authzed/api#63](https://github.com/authzed/api/pull/63) and SpiceDB itself in https://github.com/authzed/spicedb/pull/1326 and https://github.com/authzed/spicedb/pull/1281

### CODE REVIEW

File: internal/cmd/restore.go [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+
+	relationshipWriter, err := client.BulkImportRelationships(ctx)
+	if err != nil {
+		return fmt.Errorf("error creating writer stream: %w", err)
+	}
+
+	batchSize := cobrautil.MustGetInt(cmd, "batch-size")
+	batchesPerTransaction := cobrautil.MustGetInt(cmd, "batches-per-transaction")
+
+	batch := make([]*v1.Relationship, 0, batchSize)
```
Comment (jzelinskie) [inline internal/cmd/restore.go:71]: This batching logic looks good, but I think SpiceDB has a generic batcher that might be more readable. I'm not sure if it's worth pulling in.

**Review by jzelinskie — CHANGES_REQUESTED** [review]

File: internal/backupformat/decoder.go [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+	return d.schema
+}
+
+func (d *Decoder) Close() error {
+	return nil
+}
+
+func (d *Decoder) Next() (*v1.Relationship, error) {
+	if !d.dec.HasNext() {
+		return nil, nil
```
[the remark this answers is not in the document]
Comment (jakedt) [reply internal/backupformat/decoder.go:53]: I've explored that in the past, the reasons I can see that `database/sql` does that is: 1) caller managed memory control, which I'm not concerned about for a client-side tool, and 2) the APIs don't know the types of the things they're returning, so it's `...any`, we know for sure that it's a relationship.

File: internal/backupformat/encoder.go [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+	var toEncode RelationshipV1
+
+	toEncode.ObjectType = rel.Resource.ObjectType
+	toEncode.ObjectID = rel.Resource.ObjectId
+	toEncode.Relation = rel.Relation
+	toEncode.SubjectObjectType = rel.Subject.Object.ObjectType
+	toEncode.SubjectObjectID = rel.Subject.Object.ObjectId
+	toEncode.SubjectRelation = rel.Subject.OptionalRelation
+	if rel.OptionalCaveat != nil {
+		contextBytes, err := proto.Marshal(rel.OptionalCaveat.Context)
```
[the remark this answers is not in the document]
Comment (jakedt) [reply internal/backupformat/encoder.go:46]: I think I was using an old version of the protos that didn't have that or something. It's there now though!

File: internal/cmd/backup.go [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+		return fmt.Errorf("error closing backup encoder: %w", err)
+	}
+
+	if err := f.Sync(); err != nil {
+		return fmt.Errorf("error syncing backup file: %w", err)
+	}
+
+	if err := f.Close(); err != nil {
+		return fmt.Errorf("error closing backup file: %w", err)
+	}
```
[the remark this answers is not in the document]
Comment (jakedt) [reply internal/cmd/backup.go:102]: I don't think so, you can't really recover a partially created backup, and the process is about to exit anyway.

File: internal/cmd/restore.go (hunk 1 above)
Comment (jakedt) [reply internal/cmd/restore.go:71]: Does it work when you don't have everything in a slice already to start?

File: internal/backupformat/encoder.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (jakedt) [reply internal/backupformat/encoder.go:46]: Actually, the structpb impl from Google doesn't include MarshalVT. This has to use the generic one.

[the remark this answers is not in the document]
Comment (jzelinskie) [reply internal/backupformat/encoder.go:46]: Yeah that makes sense, but is also unfortunate.

File: internal/cmd/restore.go (hunk 1 above)
Comment (jzelinskie) [reply internal/cmd/restore.go:71]: Nope -- https://github.com/authzed/spicedb/blob/2ead4675d99e9a2a41dfa6f536b5079c9c626032/pkg/util/chunking.go#L7

File: internal/cmd/backup.go [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+			}
+			stored++
+
+			if stored%100_000 == 0 {
+				log.Trace().Uint("relationships", stored).Msg("progress")
+			}
+		}
+	}
+
+	log.Info().Uint("relationships", stored).Msg("finished backup")
```
[the remark this answers is not in the document]
Comment (jakedt) [reply internal/cmd/backup.go]: Discussed offline: its too hard to do this without going all in on bubbletea, maybe that could be a zed interactive-mode revamp at some point.

**Review by jzelinskie — APPROVED** [review]

**Pull request merged**